src/cupbearer/models/transformers_hf.py

```diff
@@ -0,0 +1,189 @@
+        transformer = model.transformer
+        tokenizer.pad_token = tokenizer.eos_token
+        emb_dim = transformer.embed_dim
+        max_len = tokenizer.model_max_length
+    elif name in NEOX_MODELS:
+        checkpoint = NEOX_MODELS[name]
+        model: GPTNeoXForCausalLM = AutoModelForCausalLM.from_pretrained(checkpoint)
+        tokenizer: GPTNeoXTokenizerFast = AutoTokenizer.from_pretrained(checkpoint)
+        transformer = model.gpt_neox
+        tokenizer.pad_token = tokenizer.eos_token
+        config: GPTNeoXConfig = transformer.config
+        emb_dim = config.hidden_size
+        max_len = config.max_position_embeddings
+    else:
+        raise ValueError(f"unsupported model {name}")
+
+    return transformer, tokenizer, emb_dim, max_len
+
+
+class TransformerBaseHF(HookedModel):
+    def __init__(
+        self,
+        model: PreTrainedModel,
+        tokenizer: PreTrainedTokenizerBase,
+        embed_dim: int,
+        max_length: int,  # TODO: find attribute in model,
+    ):
+        super().__init__()
+        self.model = model
+        self.tokenizer = tokenizer
+        self.max_length = max_length
+        self.embed_dim = embed_dim
+
+        # setup
+        self.tokenizer.pad_token = tokenizer.eos_token
+
+    @property
+    def defeault_names(self) -> list[str]:
+        return ["final_layer_embeddings"]
+
+    def process_input(self, x) -> TokenDict:
+        tokens = self.tokenizer(
+            x, max_length=self.max_length, padding="max_length", return_tensors="pt"
+        )
+        # load to device
+        tokens = {k: v.to(self.model.device) for k, v in tokens.items()}
+        return tokens
+
+    def get_single_token(self, x):
+        tokens: TokenDict = self.tokenizer(x)
+        return tokens["input_ids"][0]
+
+    def get_embeddings(self, tokens: TokenDict) -> torch.Tensor:
+        b, s = tokens["input_ids"].shape
+        out: BaseModelOutputWithPast = self.model(**tokens)
+        embeddings = out.last_hidden_state
+        assert embeddings.shape == (b, s, self.embed_dim), embeddings.shape
+        self.store("last_hidden_state", embeddings)
+        return embeddings
+
+
+# TODO: test
+class ClassifierTransformerHF(TransformerBaseHF):
+    def __init__(
+        self,
+        model: PreTrainedModel,
+        tokenizer: PreTrainedTokenizerBase,
+        embed_dim: int,
+        max_length: int,
+        num_classes: int,
+    ):
+        super().__init__(
+            model=model, tokenizer=tokenizer, embed_dim=embed_dim, max_length=max_length
+        )
+        self.num_classes = num_classes
+        self.classifier = nn.Linear(self.embed_dim, self.num_classes)
+
+    def forward(self, x: str | list[str]):
+        # get tokens
+        tokens = self.process_input(x)
+        # get embeddings
+        embeddings = self.get_embeddings(tokens)
+
+        # TODO (odk) se store (doesn't this slow down training?)
```

Comment:
> (doesn't this slow down training?)

Calling `self.store()` should be ~free outside a `self.capture()` context manager

Reply:
(but also like we discussed, we should switch to pytorch hooks anyway, so it won't be each model's responsibility anymore to expose activations)